File: src/core/project-detector.ts
/**
 * Project detector module for ubuild
 *
 * Detects and validates Unreal Engine projects from .uproject files.
 * Provides methods to find, parse, and validate project configurations
 * including modules, plugins, and target files.
 *
 * @module core/project-detector
 */

import fs from 'fs-extra';
import path from 'path';
import { glob } from 'glob';
import {
  UProject,
  ProjectInfo,
  ProjectDetectionOptions,
  ProjectDetectionResult,
} from '../types/project';
import { formatError } from '../utils/error';
import { inferTargetType } from '../utils/target-helpers';

/**
 * Detects and validates Unreal Engine projects from .uproject files.
 * Provides methods to find, parse, and validate project configurations.
 */
export class ProjectDetector {
  /**
   * Detects an Unreal Engine project in the specified directory.
   * @param options - Detection options including cwd and recursive search flag
   * @returns Promise resolving to project detection result with validation status
   */
  static async detectProject(
    options: ProjectDetectionOptions = {}
  ): Promise<ProjectDetectionResult> {
    const cwd = options.cwd || process.cwd();
    const warnings: string[] = [];

    try {
      const uprojectFiles = await this.findUProjectFiles(cwd, options.recursive);

      if (uprojectFiles.length === 0) {
        return {
          isValid: false,
          error: 'No Unreal Engine project (.uproject) file found',
          warnings,
        };
      }

      const uprojectPath = uprojectFiles[0];
      const projectDir = path.dirname(uprojectPath);

      const uproject = await this.parseUProject(uprojectPath);

      const validationResult = this.validateUProject(uproject);
      if (!validationResult.isValid) {
        return {
          isValid: false,
          error: `Invalid .uproject file: ${validationResult.error}`,
          warnings: [...warnings, ...validationResult.warnings],
        };
      }

      // Merge validation warnings (e.g., unexpected FileVersion) into the result
      warnings.push(...validationResult.warnings);

      const projectName = path.basename(uprojectPath, '.uproject');

      const sourceDir = path.join(projectDir, 'Source');
      const hasSourceDir = await fs.pathExists(sourceDir);

      if (!hasSourceDir) {
        warnings.push('Source directory not found - this may be a blueprint-only project');
      }

      const targets = hasSourceDir ? await this.findTargetFiles(sourceDir) : [];

      const modules = hasSourceDir ? await this.findModuleFiles(sourceDir) : [];

      const projectInfo: ProjectInfo = {
        name: projectName,
        path: projectDir,
        uproject,
        sourceDir: hasSourceDir ? sourceDir : '',
        targets,
        modules,
      };

      return {
        isValid: true,
        project: projectInfo,
        warnings,
      };
    } catch (error) {
      return {
        isValid: false,
        error: formatError(error),
        warnings,
      };
    }
  }

  /**
   * Finds .uproject files in the specified directory.
   * @param cwd - The directory to search in
   * @param recursive - Whether to search recursively in subdirectories
   * @returns Promise resolving to array of absolute paths to .uproject files
   */
  private static async findUProjectFiles(cwd: string, recursive?: boolean): Promise<string[]> {
    const pattern = recursive ? '**/*.uproject' : '*.uproject';
    return await glob(pattern, { cwd, absolute: true });
  }

  /**
   * Parses a .uproject file and returns its contents as a UProject object.
   * @param uprojectPath - Absolute path to the .uproject file
   * @returns Promise resolving to the parsed UProject object
   * @throws Error if the file cannot be read or contains invalid JSON
   */
  private static async parseUProject(uprojectPath: string): Promise<UProject> {
    const content = await fs.readFile(uprojectPath, 'utf-8');
    try {
      return JSON.parse(content);
    } catch (parseError) {
      throw new Error(`Invalid JSON in .uproject file: ${formatError(parseError)}`);
    }
  }

  /**
   * Validates that a parsed UProject object has all required fields.
   * @param uproject - The UProject object to validate
   * @returns Validation result with isValid flag, optional error message, and warnings array
   */
  private static validateUProject(uproject: UProject): {
    isValid: boolean;
    error?: string;
    warnings: string[];
  } {
    const warnings: string[] = [];

    if (uproject.FileVersion === undefined) {
      return { isValid: false, error: 'Missing FileVersion field', warnings };
    }

    if (uproject.FileVersion !== 3) {
      warnings.push(`Unexpected FileVersion: ${uproject.FileVersion}. Expected 3.`);
    }

    if (!uproject.EngineAssociation) {
      return { isValid: false, error: 'Missing EngineAssociation field', warnings };
    }

    if (!uproject.Modules || !Array.isArray(uproject.Modules)) {
      return { isValid: false, error: 'Missing or invalid Modules array', warnings };
    }

    return { isValid: true, warnings };
  }

  /**
   * Finds .Target.cs files in the Source directory and extracts target information.
   * @param sourceDir - Absolute path to the Source directory
   * @returns Promise resolving to array of target objects with name, type, and path
   */
  private static async findTargetFiles(sourceDir: string): Promise<ProjectInfo['targets']> {
    const files = await glob('*.Target.cs', { cwd: sourceDir, absolute: false });

    const targets = files.map((file) => {
      const fileName = path.basename(file, '.Target.cs');
      const type = inferTargetType(fileName);

      return {
        name: fileName,
        type,
        path: path.join(sourceDir, file),
      };
    });

    return targets;
  }

  /**
   * Finds .Build.cs files in the Source directory and extracts module information.
   * @param sourceDir - Absolute path to the Source directory
   * @returns Promise resolving to array of module objects with name and path
   */
  private static async findModuleFiles(sourceDir: string): Promise<ProjectInfo['modules']> {
    const files = await glob('**/*.Build.cs', { cwd: sourceDir, absolute: false });

    const modules = files.map((file) => {
      const moduleName = path.basename(file, '.Build.cs');
      return {
        name: moduleName,
        path: path.join(sourceDir, file),
      };
    });

    return modules;
  }
}

File: src/core/project_detector.rs
use std::path::{Path, PathBuf};

use anyhow::Result;

use crate::types::{ModuleInfo, ProjectDetectionResult, ProjectInfo, ResolvedTarget};
use crate::utils::version::infer_target_type;

use super::project_path_resolver::ProjectPathResolver;

pub struct ProjectDetector;

impl ProjectDetector {
    pub fn detect(cwd: Option<&str>, recursive: bool) -> Result<ProjectDetectionResult> {
        let dir = cwd.map_or_else(
            || std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")),
            PathBuf::from,
        );

        let mut warnings = Vec::new();

        let pattern = if recursive {
            format!("{}/**/*.uproject", dir.display())
        } else {
            format!("{}/*.uproject", dir.display())
        };

        let uproject_files: Vec<PathBuf> = glob::glob(&pattern)
            .into_iter()
            .flatten()
            .filter_map(Result::ok)
            .collect();

        let uproject_path = match uproject_files.first() {
            Some(p) => p.clone(),
            None => {
                return Ok(ProjectDetectionResult {
                    project: None,
                    warnings: vec!["No .uproject file found".to_string()],
                });
            }
        };

        let uproject = ProjectPathResolver::read_uproject(&uproject_path)?;

        if uproject.file_version != Some(3) {
            warnings.push(format!(
                "Unexpected FileVersion: {:?} (expected 3)",
                uproject.file_version
            ));
        }
        if uproject.engine_association.is_empty() {
            warnings.push("No EngineAssociation set in .uproject".to_string());
        }

        let project_name = ProjectPathResolver::project_name(&uproject_path);
        let project_dir = ProjectPathResolver::project_dir(&uproject_path);
        let source_dir = project_dir.join("Source");
        let has_source = source_dir.is_dir();

        let targets = if has_source {
            Self::find_targets(&source_dir)
        } else {
            Vec::new()
        };

        let modules = if has_source {
            Self::find_modules(&source_dir)
        } else {
            Vec::new()
        };

        Ok(ProjectDetectionResult {
            project: Some(ProjectInfo {
                name: project_name,
                path: uproject_path,
                uproject,
                source_dir: has_source.then_some(source_dir),
                targets,
                modules,
            }),
            warnings,
        })
    }

    fn find_targets(source_dir: &Path) -> Vec<ResolvedTarget> {
        let pattern = format!("{}/*.Target.cs", source_dir.display());
        glob::glob(&pattern)
            .into_iter()
            .flatten()
            .filter_map(Result::ok)
            .filter_map(|p| {
                let stem = p.file_stem()?.to_string_lossy().to_string();
                let name = stem.strip_suffix(".Target")?;
                Some(ResolvedTarget {
                    name: name.to_string(),
                    target_type: infer_target_type(name).to_string(),
                })
            })
            .collect()
    }

    fn find_modules(source_dir: &Path) -> Vec<ModuleInfo> {
        let pattern = format!("{}/**/*.Build.cs", source_dir.display());
        glob::glob(&pattern)
            .into_iter()
            .flatten()
            .filter_map(Result::ok)
            .filter_map(|p| {
                let stem = p.file_stem()?.to_string_lossy().to_string();
                let name = stem.strip_suffix(".Build")?;
                Some(ModuleInfo {
                    name: name.to_string(),
                    path: p,
                })
            })
            .collect()
    }
}
